src/facade/dragonfly_connection.cc

```diff
@@ -102,18 +102,83 @@ struct Connection::Shutdown {
 };
 
 struct Connection::Request {
-  absl::FixedArray<MutableSlice, 6> args;
+  struct PipelineMsg {
+    absl::FixedArray<MutableSlice, 6> args;
 
-  // I do not use mi_heap_t explicitly but mi_stl_allocator at the end does the same job
-  // of using the thread's heap.
-  // The capacity is chosen so that we allocate a fully utilized (256 bytes) block.
-  absl::FixedArray<char, kReqStorageSize, mi_stl_allocator<char>> storage;
-  AsyncMsg* async_msg = nullptr;  // allocated and released via mi_malloc.
+    // I do not use mi_heap_t explicitly but mi_stl_allocator at the end does the same job
+    // of using the thread's heap.
+    // The capacity is chosen so that we allocate a fully utilized (256 bytes) block.
+    absl::FixedArray<char, kReqStorageSize, mi_stl_allocator<char>> storage;
 
-  Request(size_t nargs, size_t capacity) : args(nargs), storage(capacity) {
+    PipelineMsg(size_t nargs, size_t capacity) : args(nargs), storage(capacity) {
+    }
+  };
+
+  struct PubSubMsg {
+    AsyncMsg* async_msg = nullptr;  // allocated and released via mi_malloc.
+
+    PubSubMsg(const PubMessage& pub_msg, fibers_ext::BlockingCounter bc)
+        : async_msg{(AsyncMsg*)mi_malloc(sizeof(AsyncMsg))} {
+      async_msg->bc = std::move(bc);
+      async_msg->pub_msg = pub_msg;
+    }
+
+    void Free() {
+      async_msg->bc.Dec();
+      async_msg->~AsyncMsg();
+      mi_free(async_msg);
+    }
+  };
+
+ private:
+  using MessagePayload = std::variant<PipelineMsg, PubSubMsg>;
+
+  Request(PipelineMsg msg) : payload(std::move(msg)) {
+  }
+
+  Request(PubSubMsg msg) : payload(std::move(msg)) {
   }
 
   Request(const Request&) = delete;
+
+ public:
+  static void Free(Request* self) {
```

Comment:
I prefer to pull out function implementation that takes more than 2 lines out of class declaration.

Reply:
see my comment above - if you already wrap everything into a dedicated class, the variant d'tor should do the job imho.

Reply:
yes - this was mostly as POC, in the code that support the monitor I dont have this in the class. This is mostly POC.